### PR TITLE
Refactor with NEAR standard NEP141 

### PR DESCRIPTION
**Pull request by mrLSD**
## Overview
To significantly simplify the contract, as well as to avoid potential errors and pitfalls associated with the implementation of the NEP-141, the contract is being refactored with a library `near-standard` dependency.

## What is removed
All implementations related to Fungible TOkens. And introduced dependency from `near-standard`.

## Breaking changes
NEP-141 basic (`ft_transfer`, `ft_transfer_call`) methods are now public, without access right control.
All related methods were also removed.

All types, that were related to NEP-141 implementation are deleted.

## Tests
Tests refactored qith new types dependencies and with keep in mind of new changes

## Gas cost
Gas cost increase x2 (twice) for migrations. It relates to a new organization of storage key.

### CODE REVIEW

File: eth-connector-tests/src/migration.rs [hunk 1]
```diff
@@ -111,26 +107,26 @@ async fn test_migration_state() -> anyhow::Result<()> {
         }
     }
     assert_eq!(proofs_count, data.proofs.len());
-    assert!(proofs_gas_burnt as f64 / 1_000_000_000_000. < 5416.1);
+    // INCREASED!
+    //assert!(proofs_gas_burnt as f64 / 1_000_000_000_000. < 5416.1);
```
Comment (birchmd) [inline eth-connector-tests/src/migration.rs:111]: Could you update this with the new amount of gas burnt?

**Review by birchmd — COMMENTED** [review]
Looks pretty good to me. A net decrease of almost 1500 lines of code should be good for us in the long run.

File: eth-connector-tests/src/migration.rs [hunk 2]
```diff
@@ -154,15 +150,14 @@ async fn test_migration_state() -> anyhow::Result<()> {
         accounts = HashMap::new();
     }
     assert_eq!(data.accounts.len(), accounts_count);
-    assert!(accounts_gas_burnt as f64 / 1_000_000_000_000. < 1457.);
+    // INCREASED!
+    //assert!(accounts_gas_burnt as f64 / 1_000_000_000_000. < 1457.);
```
[the remark this answers is not in the document]
Comment (mrLSD) [reply eth-connector-tests/src/migration.rs:155]: Done

File: eth-connector/src/connector_impl.rs [hunk 3]
```diff
@@ -180,3 +179,156 @@ impl ConnectorDeposit for EthConnector {
             )
     }
 }
+
+pub mod error {
```
[the remark this answers is not in the document]
Comment (mrLSD) [reply eth-connector/src/connector_impl.rs:183]: `eth-connector/src/errors.rs` includes mainly error constants and only one (Proof) error definition. It means that module for "common" use case.

In turn, the specified error submodule contains definitions of error types specific to this module. I see it as logical to separate it into a separate sub-module, not as part of the general error types.

**Comment by mrLSD** [comment]
@birchmd unfortunately, gas costs for migration increased twice (!). AFAIK it depends on how currently storage key is organized. Specifically `proof keys`.

**Comment by birchmd** [comment]
A factor of 2 is a large increase. It's not clear to me why this increase happens. You say it's because of the proof keys, but it looks like both before and after this change they were contained in a `LookupMap<String, bool>`. Could you elaborate more on what difference causes the gas increase?

Also, is this factor of 2 increase a problem for this migration?

**Comment by mrLSD** [comment]
> Could you elaborate more on what difference causes the gas increase?
Even with increased gas, all migration costs only 1.5 NEAR (in `testnet`).

What was changed: before refactoring, proof keys were part of `FungibleToken` struct. And yes it was `LookupMap`. For some reason, after moving it to the root of the Contract struct - gas costs increased.

**Comment by mrLSD** [comment]
> Are you compiling the contract with link-time optimization enabled?

Yes, it is: 
https://github.com/aurora-is-near/aurora-eth-connector/blob/feat/use-near-standard-nep141-implementation/Cargo.toml

**Review by birchmd — APPROVED** [review]
It's a little concerning that we don't understand why this changes makes the migration much more expensive. But since the migration is still possible (and relatively fast), I think we should still accept this change since it makes the maintenance burden of this code lower.

**Comment by joshuajbouw** [comment]
@birchmd @mrLSD gas cost possibly may have increased with Workspace 0.7.0?

**Comment by joshuajbouw** [comment]
Tests failing :(, please fix.

**Review by joshuajbouw — APPROVED** [review]
That was a larger PR than I expected. The tests are still failing, that really needs to get fixed. Large gas costs may have measured higher in workspaces perhaps? Or, NEAR's contract impl have a bit more checks in it. I don't really see what could've caused it but generally, if there are gas cost increases it is quite important to find out why.

**Comment by mrLSD** [comment]
> @birchmd @mrLSD gas cost possibly may have increased with Workspace 0.7.0?

It was changed after the proposal in: https://github.com/near/workspaces-rs/issues/255 ,
https://github.com/near/workspaces-rs/issues/253 

Opened 2 issues about that. Currently no success.

**Comment by joshuajbouw** [comment]
Whats the status on this?

**Pull request merged**